### PR TITLE
[fix bug] logits's shape different from label's shape in preprocess_logits_for_metrics

### PR DESCRIPTION
**Pull request by wiserxin**
# What does this PR do?

After update my tranformers  4.39.3 --> 4.41.2, training LLM with deepspeed DDP method,  something wrong when it goes to my `preprocess_logits_for_metrics` function: **logits's shape  is different from label's shape**.

Then i discover that in latest trainer [code](https://github.com/huggingface/transformers/blob/377e90392804790d76d818264d39c694a71e07e0/src/transformers/trainer.py#L3811), before passing to preprocess_logits_for_metrics,  logits are padded by function `pad_across_processes` but labels are not. 

It seems like a bug unexpectedly introduced .

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @muellerzr @SunMarc

**Comment by SunMarc** [comment]
Hi @wiserxin, this bug was indeed introduced by this PR https://github.com/huggingface/transformers/pull/30267. The padding of the labels happens after `preprocess_logits_for_metrics`, hence the issue that you are experiencing. @qubvel, is it safe to put it back where it was ?

**Comment by qubvel** [comment]
Oh, that's true. @SunMarc yes, should be safe. Alternatively, we can just move the "if" block with labels padding before the "if" block with logits.

**Comment by SunMarc** [comment]
> Alternatively, we can just move the "if" block with labels padding before the "if" block with logits

Yeah, I think that's would be better for readability + it was the previous behavior ! Could you make the change @wiserxin ?

**Comment by wiserxin** [comment]
> > Alternatively, we can just move the "if" block with labels padding before the "if" block with logits
> 
> Yeah, I think that's would be better for readability + it was the previous behavior ! Could you make the change @wiserxin ?

sure

**Review by SunMarc — APPROVED** [review]
LGTM ! Thanks for the PR @wiserxin !

**Comment by qubvel** [comment]
@SunMarc @wiserxin let's maybe add a comment about the order of this "if" block to avoid the same bug in the future. :)

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

+1 on @qubvel's comment

**Comment by wiserxin** [comment]
> @SunMarc @wiserxin let's maybe add a comment about the order of this "if" block to avoid the same bug in the future. :)

good idea

**Review by muellerzr — APPROVED** [review]
Thanks a bunch!
